src/lib.rs

```diff
@@ -398,9 +400,32 @@ pub async fn call<'a, T: hyper::client::connect::Connect + Clone + Send + Sync +
         client_ip
     );
 
+    let request_upgraded = request.extensions_mut().remove::<OnUpgrade>();
+
     let proxied_request = create_proxied_request(client_ip, forward_uri, request)?;
+    let mut response = client.request(proxied_request).await?;
+
+    if response.status() == StatusCode::SWITCHING_PROTOCOLS {
+        let mut response_upgraded = response
+            .extensions_mut()
+            .remove::<OnUpgrade>()
+            .expect("response does not have an upgrade extension")
+            .await?;
+
+        tokio::spawn(async move {
+            let mut request_upgraded = request_upgraded
+                .expect("request does not have an upgrade extension")
+                .await
+                .expect("failed to upgrade request");
+
+            copy_bidirectional(&mut response_upgraded, &mut request_upgraded)
+                .await
+                .expect("coping between upgraded connections failed");
+        });
+
+        return Ok(response);
```

Comment:
After working with the code some more I have a couple more comments:

1. This return statement makes the flow slightly confusing. It would be more idiomatic if you directly returned the if statement, with an else statement for the other branch.
2. Whenever the response status is SWITCHING PROTOCOLS, your code doesn't strip any of the hop-by-hop headers as well as the connection headers. This doesn't seem like it's correct, isn't it just the Connection and Upgrade headers which should remain with the websocket specific values?

Reply:
I'm working on an update for 1 and the previous comment.

For 2, it seems the [go code branches](https://github.com/golang/go/blob/0668e3cb1a8407547f1b4e316748d3b898564f8e/src/net/http/httputil/reverseproxy.go#L321) into `handleUpgradeResponse()` just before removing the connection headers and hop header. I also don't see it touching the hop headers inside the `handleUpgradedResponse()` method. But I could easily have missed this as I'm not familiar with `Hijacker` :see_no_evil:.